# 01. Import Libraries

In [1]:
#Import libraries
import pandas as pd
import numpy as np
import os

# 02. Import Data

In [2]:
#Folder Path
path = r"C:\Users\marce\Documents\Instacart Basket Analysis"

In [3]:
# Import products.csv
df_prods= pd.read_csv(os.path.join (path, '02 Data', 'Original Data', 'products.csv'), index_col= False)

In [4]:
# Import orders_wrangled.csv
df_ords=pd.read_csv(os.path.join (path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# 03. Consistency Check df_prods

In [42]:
#Descriptive Statistics
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [43]:
# Max for prices of 99999 seems to be an error , as it deviates a lot from the other prices if we compare the quartiles

In [25]:
df_prods[df_prods.prices == 99999]

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0

In [26]:
# Looking for item by Name
df_prods[df_prods.product_name== 'Reduced Fat Milk']

product_id      product_name  aisle_id  department_id  prices
5612        5612  Reduced Fat Milk        84             16     9.2

## Mixed Data Types

In [6]:
# Create a dataframe

df_test = pd.DataFrame()

In [7]:
# Create a mixed  column type

df_test['mix'] = ['a', 'b', 1, True]


In [8]:
# Display first 5 rows
df_test.head()

mix
0     a
1     b
2     1
3  True

In [9]:
# Check for mixed data  types

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)
    

mix


In [10]:
#Change the column data type

df_test['mix'] = df_test['mix'].astype('str')

In [48]:
# Check for mixed-type columns in df_prods
for col in df_prods.columns.tolist():
    weird = (df_prods[[col]].applymap(type) != df_prods[[col]].iloc[0].apply(type)).any(axis =1)
    if len (df_prods[weird])>0:
        print(col)

product_name


In [ ]:
# 'product_name' column contains mixed -type data. Will be addressed after checking the missing values.

# Missing Values

In [11]:
# Count the missing values in each column:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [12]:
# Create a new data frame containing only 'product name' values that meet the condition isnull() =True

df_nan = df_prods[df_prods['product_name'].isnull()==True]

In [13]:
# Display df_nan
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [14]:
df_prods.shape

(49693, 5)

In [15]:
#Create new dataframe without missing values

df_prods_clean = df_prods[df_prods['product_name'].isnull()==False]


In [16]:
df_prods_clean.shape

(49677, 5)

# Duplicates

In [17]:
# Create new subset of df_prods_clean containing only duplicate rows

df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [18]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [19]:
df_prods_clean.shape

(49677, 5)

In [20]:
# create new dataframe without duplicates

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [21]:
df_prods_clean_no_dups.shape

(49672, 5)

# Task 4.5 

#  2 Run the df.describe() function on your df_ords dataframe.
Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.

In [44]:
#Descriptive Statistics for df_ords
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [24]:
#  Descriptive Statistics for df_ords seems to be normal

#  3 Check for mixed-type data in  df_ords dataframe

In [29]:
# Check for mixed-type data

for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type)!= df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird])>0:
        print(col)

In [31]:
 # Step 4  No mixed - type data was detected for df_ords

#  5 Run a check for missing values in df_ords dataframe

In [32]:
# Check for missing values in df_ords

df_ords.isnull().sum()


Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [33]:
# Create a new data frame containing missing values

df_ords_nan= df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [34]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [35]:
df_ords

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

#The missing data in the days_since_prior_order column could represent the first customer orders, since the missing data 
value of 206209 coincides with the last user_id in the new dataframe df_ords_nan.From df_ords we notice that there are a
total of 206209 user_ids and 3421083 orders

# 6. Adress the missing values

#In this case, I believe that the missing values should not be subject to changes, because they do not affect our analysis
and "1" from order_number could be considered as a flag for the first order

# 7. Check for duplicate values in df_ords data

In [39]:
# Create a new subset with duplicates

df_ords_dups = df_ords[df_ords.duplicated()]

In [40]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [41]:
# No duplicates detected

# 9. Export final cleaned df_prods and df_ords data 
as 'csv' files in 'Prepared Data' folder

In [49]:
# Export df_prods_clean_no_dups
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

In [50]:
# Export final df_ords
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))